### ISSUE TITLE
Docker rootless network performance enhancements

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#12]
There's a more efficient way to handle networking than what it is now essentially; "netcat" piping through data.
Current method limits the ultimate performance for end user

'lxc-user-nic' or something.
Should benchmark which is fastest // most efficient method of accomplish this.

routed is an option too, but no per user ipv4 addresses (gets expensive, fast)

**Comment by MagnaCapax** [comment]
## Adversarial Review — Follow-Up

**Issue:** #12 — Docker Rootless Network Performance Enhancements
**Reviewed:** 2026-04-16
**Verdict:** SHOULD-NOT-IMPLEMENT (prior verdict stands — additional evidence below)

> Prior review (2026-04-08) reached the correct verdict. This follow-up adds codebase findings from direct source inspection that were not in that review. No new comments have appeared since. The prior findings are not repeated here.

---

### New Finding: The "NOT VIABLE" Verdict Is Now Encoded in Documentation

`docs/install-media-stack.md` explicitly documents, as shipped code:

> "Docker rootless mode allows running containers without root privileges. While this sounds appealing for shared hosting, testing has shown significant issues on PMSS infrastructure."
> 1. **cgroup Delegation Issues**: Proxmox VMs have limited cgroup v2 delegation, causing containers to fail with "Permission denied" errors during startup.
> 2. **Higher Resource Usage**: Docker daemon alone uses ~180 MB.
> 3. **System Configuration Required**: Docker rootless requires root-level setup that shared hosting users cannot perform.

This is not just a research note in issue #171 — it is the documented rationale in the installed media stack documentation. Any implementation of #12 (network backend replacement) must first address the premise this documentation records. The codebase itself has moved past the point where networking is the binding constraint.

---

### Codebase State (Source-Verified, This Review)

**Current network stack in code:**

- `scripts/lib/update/apps/packages.php` queues: `slirp4netns`, `uidmap`, `fuse-overlayfs` for all installations
- `scripts/lib/update/apps/docker.php`: pins slirp4netns to **v1.3.2** via GitHub download for Debian <12; Debian 12 uses the apt-packaged version from the dpkg baseline (`scripts/lib/update/dpkg/selections-debian12.txt`)
- `scripts/lib/update/distUpgrade.php` lines 202–206: dist-upgrade cleanup removes the conflicting `/usr/local/bin/slirp4netns` when Debian 12's apt version is present — showing the version-management debt is already partially acknowledged in code
- `scripts/util/userDocker.php`: 415-line helper managing per-user rootless Docker lifecycle; the dominant mode is explicitly `dockerd-rootless.sh` (non-systemd path); cgroup v2 handling writes `cgroupdriver=cgroupfs` to `daemon.json` as a workaround for the delegation failure

**`lxc-user-nic` and `pasta`/`passt`:** Zero occurrences anywhere in the codebase — no package install, no config, no documentation, no ADR, no test. These candidates are purely notional.

**`setupNetwork.php`** exists at `scripts/util/setupNetwork.php` but contains no Docker-specific logic — it is unrelated to container networking.

---

### Supersession Assessment

The prior review cited #171 as a blocker. Since then, the NOT VIABLE finding from #171 has been incorporated into `docs/install-media-stack.md` as shipped documentation. The codebase's direction is native installation (Servarr, Jellyfin via .NET binary) rather than containerisation for the shared hosting tier. Network backend optimisation for rootless Docker is an improvement to a path the codebase is documenting as non-viable for the primary deployment target.

The one bounded change that remains valid (bumping the slirp4netns version pin in `docker.php` for Debian <12) is a maintenance task independent of this issue's scope.

---

### Anti-Pattern Update (Source-Verified)

| Pattern | Result | Source evidence |
|---|---|---|
| `PREMATURE_OPTIMIZATION` | **FAIL** | Measured baseline on PMSS hardware absent; startability not confirmed resolved fleet-wide per `docs/install-media-stack.md` |
| `OPERATOR_HANDOFF_ANTIPATTERN` | **FAIL** | "Should benchmark which is fastest" — still no owner, method, or metrics |
| `FABRICATED_AUTHORITY_OR_PATH` | **PASS** | All paths verified against source files in this review |

---

### One Actionable Item That Can Land Independently

`scripts/lib/update/apps/docker.php` pins slirp4netns at v1.3.2 for Debian <12. The `distUpgrade.php` cleanup already handles the conflict when upgrading to Debian 12. The pin bump is bounded, safe, and has no dependency on architecture decisions about `lxc-user-nic` or `pasta`. If there is desire to make incremental progress on this issue, that is the only change with a clear, low-risk path.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Follow-Up

**Issue:** #12 — Docker Rootless Network Performance Enhancements
**Reviewed:** 2026-04-23
**Verdict:** SHOULD-NOT-IMPLEMENT (prior verdict stands — updated context below)

> Two prior reviews: 2026-04-08 (SHOULD-NOT-IMPLEMENT) and 2026-04-16 (same verdict, codebase-verified). This third review is triggered by a material operator commit since the last review that changes the direction assessment without changing the networking verdict. Prior findings are not repeated here.

---

### New Development Since 2026-04-16 Review

**Commit `21800195` (2026-04-11, operator-authored):** `fix: rootless docker — seed cgroup v2 daemon config`

This commit advances Docker rootless viability by:
- Adding `etc/skel/.config/docker/daemon.json` with `native.cgroupdriver=cgroupfs` — all new accounts get the workaround at provision time
- Adding `userDockerEnsureCgroupfsDaemonConfig()` in `scripts/util/userDocker.php` — backfills the config for existing accounts on every `start`/`restart`
- Adding a test suite (`scripts/lib/tests/development/userDockerCgroupDriverTest.php`) covering both the skel seed and the backfill path
- Documenting the behaviour in `docs/docker-help.md`

The prior review cited cgroup delegation failure as the primary container startability gate. That gate has been partially addressed. The codebase direction is no longer "native installation only" — it is "native installation for shared hosting, Docker rootless with workarounds for users who want it." `docs/install-media-stack.md` still recommends native over rootless, but the "Permission denied" cgroup failure is now mitigated by the shipped workaround.

---

### Updated Codebase State (Source-Verified, This Review)

**Networking backend — unchanged from 2026-04-16:**
- `scripts/lib/update/apps/docker.php` line 25: slirp4netns pinned to **v1.3.2** via GitHub download for Debian <12
- `scripts/lib/update/apps/packages.php` line 82: `slirp4netns` queued via apt for Debian 12+
- `scripts/lib/update/distUpgrade.php` lines 202–206: conflict cleanup on dist-upgrade to Debian 12
- `pasta`, `passt`, `lxc-user-nic`, `bypass4netns`: **zero occurrences** anywhere in the codebase

**cgroup handling — new since 2026-04-16:**
- `scripts/util/userDocker.php`: new `userDockerCgroupMode()` detects v1/v2 via `/sys/fs/cgroup/cgroup.controllers` and `/proc/self/mountinfo`; new `userDockerEnsureCgroupfsDaemonConfig()` idempotently writes `cgroupdriver=cgroupfs` to `~/.config/docker/daemon.json`; called unconditionally on every `start`/`restart`
- `etc/skel/.config/docker/daemon.json`: new file, seeds the same config at account creation

---

### Adversarial Findings

**[C] lxc-user-nic multi-tenant security analysis still absent**

Prior review flagged this as a Critical gate. Nothing has changed: `lxc-user-nic` has zero code presence, and the multi-tenant bridge ARP-spoofing/traffic-sniffing risk on a shared host is unassessed. AGENTS.md security doctrine requires attack surface analysis before implementation. This finding stands unchanged.

**[M] The bounded slirp4netns pin bump still has not landed**

Both prior reviews identified bumping `docker.php` line 25 from v1.3.2 to current stable as the one low-risk, architecture-independent improvement available. The operator's April 11 commit advanced cgroup handling but left the version pin untouched. Worth verifying whether a newer stable release actually exists before treating this as debt.

**[M] `userDockerEnsureCgroupfsDaemonConfig()` silently recreates daemon.json on every cold-start for v2 hosts**

The function is idempotent once the `exec-opts` entry exists. However, a user who deliberately removes `daemon.json` to test a different driver will have it silently recreated on the next `start`. The behaviour is documented in `docs/docker-help.md` but the log message only fires on write — nothing in `status` output signals that the override is active. Low severity; worth noting for transparency.

**[M] No ADR for the cgroupfs workaround decision**

AGENTS.md requires ADRs for significant behavioural changes. The April 11 commit silently writes into every Docker-enabled user's `~/.config/docker/` on cgroup v2 hosts. This warrants an ADR capturing the delegation failure root cause, why `cgroupfs` driver was chosen over full cgroup v2 delegation, and any known consequences.

**[L] `docs/install-media-stack.md` Docker section still lists the now-mitigated cgroup failure as a hard blocker**

The "Alternative: Docker Rootless" section lists "cgroup Delegation Issues: causing containers to fail with 'Permission denied' errors" as reason #1 against rootless. The April 11 commit specifically addresses this. The docs should be updated to reflect that PMSS now seeds and backfills the cgroupfs workaround automatically.

---

### Anti-Pattern Hunt

| Pattern | Result | Evidence |
|---|---|---|
| `REGEX_FOR_SEMANTICS` | PASS | Not applicable |
| `PREMATURE_OPTIMIZATION` | **FAIL** | No PMSS-specific networking throughput benchmark exists at torrent-client connection counts (200+ concurrent peers). April 11 commit addresses startability, not throughput. |
| `OPERATOR_HANDOFF_ANTIPATTERN` | **FAIL** | "Should benchmark which is fastest" — still no owner, method, or metrics in the issue or any related commit |
| `FALSE_COMPLETION_CLAIM` | PASS | Issue remains open; no false completion claimed |
| `FABRICATED_AUTHORITY_OR_PATH` | PASS | All paths verified against current source |
| `FAIL_OPEN_SAFETY_BEHAVIOR` | PASS | Not applicable |
| `DRY_RUN_THEATER` | PASS | Not applicable |
| `INVESTIGATION_LOOP_THEATER` | PASS | Material new commit exists; review warranted |

---

### Caller Analysis (Networking Backend)

Files that must be updated for any networking backend replacement:
- `scripts/lib/update/apps/docker.php` line 25 — downloads slirp4netns v1.3.2 binary for Debian <12
- `scripts/lib/update/apps/packages.php` line 82 — queues `slirp4netns` apt package for Debian 12+
- `scripts/lib/update/distUpgrade.php` lines 202–206 — removes `/usr/local/bin/slirp4netns` conflict on dist-upgrade

---

### Worth-Implementing Analysis

- **Cost**: High — requires PoC integration, multi-tenant security analysis (lxc-user-nic bridge risks), PMSS-specific benchmarks at representative workloads, ADR, and updates across all three caller sites above
- **Benefit**: Throughput improvement for Docker rootless users; the April 11 commit has improved startability, making networking the next meaningful bottleneck IF the Docker rootless user base is significant. No telemetry on how many accounts actually run rootless Docker on PMSS.
- **Verdict rationale**: The primary prerequisite (container startability) has been partially addressed. The issue's value proposition has marginally increased. The verdict remains SHOULD-NOT-IMPLEMENT because the three missing prerequisites — security analysis for lxc-user-nic, PMSS-specific throughput benchmark at torrent workloads, ADR — are still entirely absent.
- **Alternative**: (1) Verify whether slirp4netns v1.3.2 is behind current stable and land the bounded pin bump if so. (2) Update `docs/install-media-stack.md` Docker section to reflect that the cgroup v2 startup failure is now mitigated.

---

### Recommendations

1. Update `docs/install-media-stack.md` — cgroup delegation blocker (reason #1) is now mitigated by the April 11 commit; the section should reflect current state.
2. Write an ADR for the `cgroupdriver=cgroupfs` workaround — the April 11 commit has no ADR coverage despite being a fleet-wide behavioural change to the per-user environment.
3. Before progressing networking backend work, measure actual throughput at torrent-client workloads (200+ concurrent peer connections) on PMSS hardware — this is the missing primitive the issue has needed since it was filed.
4. If lxc-user-nic remains a candidate: complete a multi-tenant security analysis covering bridge ARP spoofing and inter-tenant traffic visibility before any PoC work.

— Sampsa Pellervoinen 🌱

**Comment by MagnaCapax** [comment]
## Adversarial Review — Closing Comment

**Issue:** #12 — Docker Rootless Network Performance Enhancements
**Reviewed:** 2026-05-23 (4th review)
**Verdict:** SHOULD-NOT-IMPLEMENT (prior verdict reconfirmed — additional alternatives surveyed below). Closing as `not_planned`. The one bounded change (slirp4netns pin bump) is documented at the bottom for a separate small-issue follow-up.

> Three prior reviews: 2026-04-08, 2026-04-16, 2026-04-23. This review surveys the alternatives the original issue did not name (`bypass4netns` in particular), verifies the v1.3.3 release of the pinned binary, and closes the issue because every alternative is either net-slower at PMSS workloads, experimental without Docker support, or has unaddressed multi-tenant security gaps.

---

### Alternative survey — every faster-than-slirp4netns option assessed

#### 1. `pasta` (passt project)
- RootlessKit network driver, ships in Docker Engine 25.0+
- **Benchmarks (Podman discussion #22559, conventional HTTP load):**
  - c=1: pasta 1699 vs slirp4netns 1148 req/s — pasta wins
  - c=8: pasta 13852 vs slirp4netns 13925 — break-even
  - c=16: pasta 14289 vs **slirp4netns 24852** — slirp +74%
  - c=32: pasta 15166 vs **slirp4netns 28931** — slirp ~2×
- **PMSS workload:** a single active torrent client makes 200+ concurrent peer connections; multi-app users much more. This is the regime where slirp4netns wins decisively.
- Podman 5.8 (March 2026) defaulted to pasta — Podman optimises for low-concurrency app workloads (web servers, CLI tools), not torrent peer concurrency.
- Verdict: net regression for PMSS.

#### 2. `bypass4netns`
- Status: **experimental**, kernel ≥5.9, runc ≥1.1 or crun ≥1.6, libseccomp ≥2.5
- ACM 2024 paper claims "30× faster throughput, as fast as `--net=host`" — measured against unoptimised slirp4netns at low concurrency
- **Docker integration: TODO.** Native integration exists only for nerdctl (`containerd-rootless-setuptool.sh install-bypass4netnsd`). For Docker, setup requires:
  1. Run `bypass4netnsd` daemon out-of-band
  2. Generate seccomp profile per `./test/seccomp.json.sh > $HOME/seccomp.json`
  3. Customer must launch every container with `docker run --security-opt seccomp=$HOME/seccomp.json --runtime=runc …`
- **PMSS impact:** there is no way to make this transparent to customers without a wrapper layer that doesn't exist. Customers running raw `docker run` (or `docker compose up`) would not get the speedup at all — they'd need to update every container/compose-file with the seccomp profile.
- Known limitations: host abstract sockets and loopback IPs (127.0.0.0/8) blocked; UDP port mappings ignored; potential TOCTOU exploitation via `struct sockaddr` pointers; port-forwarded ports inaccessible across containers.
- Verdict: experimental, no Docker integration, would require non-trivial PMSS wrapper layer, customer behaviour change, AND security review for the seccomp-notify TOCTOU surface.

#### 3. `lxc-user-nic`
- Status in PMSS: zero code (per prior reviews)
- Performance: approaches root Docker (real bridged networking)
- **Blocker unchanged from prior reviews:** multi-tenant ARP spoofing / cross-tenant traffic sniffing on shared bridge — unassessed. C-severity gate per AGENTS.md security doctrine.
- Verdict: NOT cleared for shared hosting until security analysis lands.

#### 4. IPv6-routed (the option dismissed in original issue body)
- Original body dismissed "routed" with "no per user IPv4 addresses (gets expensive, fast)" — IPv4-only framing
- Dual-stack IPv6 prefix delegation per user would eliminate NAT entirely → no user-space network stack cost
- Status: not implemented; would require per-user IPv6 allocation + Docker rootless IPv6 config + customer education
- Verdict: legitimate long-term direction, but a different scope of work (network architecture, not network backend choice)

#### 5. `gvisor-tap-vsock`
- RootlessKit experimental driver, originally for Podman-machine on macOS
- Not applicable to Linux bare-metal/VM PMSS hosts
- Verdict: N/A

---

### slirp4netns v1.3.2 → v1.3.3 verification (the one bounded item)

Downloaded both v1.3.2 and v1.3.3 static binaries from upstream and verified locally on 2026-05-23:

```
slirp4netns version 1.3.2     slirp4netns version 1.3.3
libslirp: 4.9.0               libslirp: 4.9.1
libseccomp: 2.5.3             libseccomp: 2.5.3
SLIRP_CONFIG_VERSION_MAX: 6   SLIRP_CONFIG_VERSION_MAX: 6
```

**Upstream release notes for v1.3.3 (June 2025):** *"this release is not worth upgrading, unless you are using the official statically-linked binary release."* PMSS uses the statically-linked binary for Debian <12 — so the upgrade does land the libslirp 4.9.0 → 4.9.1 security/perf fixes for us, but the diff is **only** the bundled libslirp version. No source changes in slirp4netns itself. `SLIRP_CONFIG_VERSION_MAX` unchanged → no wire-format / API change.

**Recommend:** file a small separate maintenance issue for `scripts/lib/update/apps/docker.php` line 25 (`…/v1.3.2/slirp4netns-…` → `…/v1.3.3/slirp4netns-…`). No ADR required, no dependency on this issue. This issue (#12) is the wrong scope for that bump.

---

### Why this issue is closed, not deferred

The body asks "should benchmark which is fastest/most efficient". After four reviews and an alternative survey:

1. **There is no candidate that is both (a) faster than slirp4netns at PMSS torrent-concurrency workloads AND (b) production-ready for Docker AND (c) cleared on multi-tenant security.**
2. The premise "current method limits ultimate performance" is correct only at low concurrency — at PMSS-realistic concurrency (16+ connections per app, multiple apps per user), slirp4netns is already the fastest user-space option.
3. The April 11 commit (`21800195`, cgroup v2 workaround) addressed the higher-priority gate — container startability — so networking is no longer blocked by a prerequisite, but it is also no longer the binding constraint on customer experience.
4. `docs/install-media-stack.md` still documents native installation (Servarr, Jellyfin .NET binary) as the preferred path for shared hosting. Networking optimisation on a deliberately-secondary path is not where engineering effort buys customer-visible wins.

Reopening criteria: if a future RootlessKit network driver demonstrates both (a) >slirp4netns throughput at parallelism >16 on PMSS hardware AND (b) cleared multi-tenant security model AND (c) production-ready Docker integration, this issue can be reopened with the benchmark data in the body.

— Väinämöinen
